### PR TITLE
Fail on compiler warnings

### PR DESCRIPTION
**Pull request by carterkozak**
**Goals (and why)**:

Catch bugs at compile time rather than runtime.

**Implementation Description (bullets)**:

Enable `-Werror`, and disable error-prone on generated projects (those suffixed `-protobufs`).
Warnings that do not currently pass are disabled for now, allowing current and new checks to be automatically validated.

**Testing (What was existing testing like?  What have you done to improve it?)**:
These are tests.

**Concerns (what feedback would you like?)**:

**Where should we start reviewing?**:
gradle files.

**Priority (whenever / two weeks / yesterday)**:
The sooner it merges, the less likely we are to regress.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -107,23 +107,98 @@ allprojects {
...
     tasks.withType(JavaCompile) {
+        options.compilerArgs += ['-Werror']
         // temporarily relax constraints until we can fix all violations
-        options.errorprone.warn 'CatchBlockLogException',
+        options.errorprone.disable 'AlmostJavadoc', 
+                'AnnotateFormatMethod',
+                'ArrayAsKeyOfSetOrMap',
+                'AssertEqualsArgumentOrderChecker',
+                'AssertionFailureIgnored',
+                'ByteBufferBackingArray',
```
Comment (carterkozak) [inline build.gradle:117]: Potential data correctness issues, we should fix the underlying code and enforce this asap.

File: build.gradle [hunk 2]
```diff
@@ -107,23 +107,98 @@ allprojects {
...
+                'CacheLoaderNull',
+                'CatchBlockLogException',
+                'CatchFail',
                 'CheckReturnValue',
+                'ClassCanBeStatic',
```
Comment (carterkozak) [inline build.gradle:122]: Trivial to fix (as many of these are), should be done in a follow-up.

**Review by jeremyk-91 — APPROVED** [review]
Yeah, the list of exceptions is pretty gory but in principle this is the right thing to do, and should get us on a path to avoid further regressions.

File: build.gradle [hunk 3]
```diff
@@ -31,7 +31,7 @@ buildscript {
 plugins {
     id 'com.github.johnrengelman.shadow' version '5.1.0'
     id 'com.palantir.configuration-resolver' version '0.4.0'
-    id 'com.palantir.consistent-versions' version '1.17.3'
+    id 'com.palantir.consistent-versions' version '1.26.1'
```
Comment (carterkozak) [inline build.gradle:34]: Note this upgrade reduced the gradle `CONFIGURING` time from ~50 seconds to ~4 seconds on my machine.

**Pull request merged**